iroh-net/src/lib.rs

```diff
@@ -103,8 +112,72 @@
 //! up the [`NodeId`] in the DNS system to find the addressing details.  This enables
 //! connecting using only the [`NodeId`] which is often more convenient and resilient.
 //!
+//! See [the discovery module] for more details.
+//!
+//!
+//! # Examples
+//!
+//! The central struct is the [`Endpoint`], which allows you to connect to other nodes:
+//!
+//! ```no_run
+//! use anyhow::Result;
+//! use iroh_net::{Endpoint, NodeAddr};
+//!
+//! async fn connect(addr: NodeAddr) -> Result<()> {
+//!     // The Endpoint is the central object that manages an iroh-net node.
+//!     let ep = Endpoint::builder().bind().await?;
+//!
+//!     // Establish a QUIC connection, open a bi-directional stream, exchange messages.
+//!     let conn = ep.connect(addr, b"hello-world").await?;
+//!     let (mut send_stream, mut recv_stream) = conn.open_bi().await?;
+//!     send_stream.write_all(b"hello").await?;
+//!     send_stream.finish()?;
+//!     let msg = recv_stream.read_to_end(10).await?;
+//!     println!("Server message: {}", String::from_utf8_lossy(&msg));
+//!
+//!     // Gracefully close the connection and endpoint.
+//!     conn.close(1u8.into(), b"done");
+//!     ep.close(0u8.into(), b"ep closing").await?;
+//!     println!("Client closed");
+//!     Ok(())
+//! }
+//! ```
+//!
+//! Every [`Endpoint`] can also accept connections:
+//!
+//! ```no_run
+//! use anyhow::{Context, Result};
+//! use futures_lite::StreamExt;
+//! use iroh_net::ticket::NodeTicket;
+//! use iroh_net::{Endpoint, NodeAddr};
+//!
+//! async fn accept() -> Result<()> {
+//!     // To accept connections at least one ALPN must be configured.
+//!     let ep = Endpoint::builder()
+//!         .alpns(vec![b"hello-world".to_vec()])
+//!         .bind()
+//!         .await?;
+//!
+//!     // Accept a QUIC connection, accept a bi-directional stream, exchange messages.
+//!     let conn = ep.accept().await.context("no incoming connection")?.await?;
+//!     let (mut send_stream, mut recv_stream) = conn.accept_bi().await?;
+//!     let msg = recv_stream.read_to_end(10).await?;
+//!     println!("Client message: {}", String::from_utf8_lossy(&msg));
+//!     send_stream.write_all(b"world").await?;
+//!     send_stream.finish()?;
+//!
+//!     // Wait for the client to close the connection and gracefully close the endpoint.
+//!     conn.closed().await;
+//!     ep.close(0u8.into(), b"ep closing").await?;
+//!     Ok(())
+//! }
+//! ```
```

Comment:
If we *really* want to, we can slim down the example like this:

```suggestion
//! ```no_run
//! # use anyhow::Result;
//! # use iroh_net::{Endpoint, NodeAddr};
//! #
//! # async fn connect(addr: NodeAddr) -> Result<()> {
//! // The Endpoint is the central object that manages an iroh-net node.
//! let ep = Endpoint::builder().bind().await?;
//!
//! // Establish a QUIC connection, open a bi-directional stream, exchange messages.
//! let conn = ep.connect(addr, b"hello-world").await?;
//! let (mut send_stream, mut recv_stream) = conn.open_bi().await?;
//! send_stream.write_all(b"hello").await?;
//! send_stream.finish()?;
//! let msg = recv_stream.read_to_end(10).await?;
//!
//! // Gracefully close the connection and endpoint.
//! conn.close(1u8.into(), b"done");
//! ep.close(0u8.into(), b"ep closing").await?;
//! # Ok(())
//! # }
//! ```
//!
//! Every [`Endpoint`] can also accept connections:
//!
//! ```no_run
//! # use anyhow::{Context, Result};
//! # use futures_lite::StreamExt;
//! # use iroh_net::ticket::NodeTicket;
//! # use iroh_net::{Endpoint, NodeAddr};
//! #
//! const ALPN: &[u8] = b"hello-world"; 
//! #
//! # async fn accept() -> Result<()> {
//! // To accept connections at least one ALPN must be configured.
//! let ep = Endpoint::builder().alpns(vec![ALPN.to_vec()]).bind().await?;
//!
//! // Accept a QUIC connection, accept a bi-directional stream, exchange messages.
//! let conn = ep.accept().await.context("no incoming connection")?.await?;
//! let (mut send_stream, mut recv_stream) = conn.accept_bi().await?;
//! let msg = recv_stream.read_to_end(10).await?;
//! send_stream.write_all(b"world").await?;
//! send_stream.finish()?;
//!
//! // Wait for the client to close the connection and gracefully close the endpoint.
//! conn.closed().await;
//! ep.close(0u8.into(), b"ep closing").await?;
//! # Ok(())
//! # }
//! ```
```

I'm stripping out the printlns and hiding the imports and function scaffolding here.
This might be fine:
1. Even if you can copy-paste these examples, you can't immediately run them, because there's no `main` function.
2. Because you can't run them, you don't need to have `println`s here, I think.
3. I hope that people don't need to see imports here, as they are already *in the docs* where they can search for everything.

I can see people disagreeing with me here, even I'm not sure, but being more elaborate is the easier option.